### DISCUSSION TITLE
Modulate voice client sound sources?

### DISCUSSION THREAD

**Discussion by rght925h** [vlad2305m/Sound-Physics-Fabric#11] [Ideas]
Likely a naive question but would it be possible to **modulate game-integrated players' proximity voice chat sources**?

[Plasmo Voice](https://www.curseforge.com/minecraft/mc-mods/plasmo-voice-client), which is based on [Simple Voice Chat](https://github.com/henkelmax/simple-voice-chat), comes to mind. I am not sure how the sound playback is handled there, of course, but it _is_ entirely ingame.

**Comment by thedocruby** [comment]
I came here to start a discussion on this very topic. I think it is a wonderful idea, and could greatly enhance the functionality and immersion factor of proximity voice chats.

**Reply by vlad2305m** [reply]
Is mod compatibility even a thing in Fabric?

**Reply by rght925h** [reply]
Well, I had the naive hope that Simple Voice Chat would somehow create a sound source at the player and dynamically play VOIP over that. Sound-Physics could then pick that up like any other sound that isn't blacklisted, no? Anyway, I have contacted the SVC developers on their Discord about this a while ago and they explained to me that this is not how it works. Simple Voice Chat uses an entirely "decoupled" sound system that runs on top of Minecraft. So unless they add their own sound modulation solution (so far nothing is planned to my knowledge), I do not see this working out. Still, would be nice ;-)

**Reply by thedocruby** [reply]
@rght925h I think PlasmoVoice is already compatible with SoundPhysics in 1.17, at least according to their curseforge page. @vlad2305m And yes, mod compatibility on Fabric is even more of a thing than it is on Forge

**Reply by vlad2305m** [reply]
@thedocruby Sorry, but can you send me a link to how to use compatibility? I have problems with figuring out how to use loom/gradle/maven to get the target mods into the environment. Thank you!

**Reply by thedocruby** [reply]
@vlad2305m I don't do much modding, but i'm well versed in combing through documentation when i don't know how to do something, so I'm happy to help. First, from the mixin docs:

> A Mixin marked as `@Pseudo` is allowed to target classes which are not available at compile time and may not be available at runtime. This can be used in situations where - for one reason or another - a target class is not available to a project being compiled and would therefore fail AP verification. The target class is simulated by the AP using knowledge of only the superclass of the target (or at least, any _available_ known superclass of the target.)

Taken from [this Mixin docs page](https://jenkins.liteloader.com/view/Other/job/Mixin/javadoc/org/spongepowered/asm/mixin/Pseudo.html)

Also, more fabric-specific, and useful when you don't need to mess with mixins, is the FabricLoader#isModLoaded() method:

> `boolean isModLoaded​(String id)`
Checks if a mod with a given ID is loaded.
Parameters:
`id` - the ID of the mod, as defined in `fabric.mod.json`
Returns:
whether or not the mod is present in this Fabric Loader instance

Taken from [this Fabris Docs page](https://maven.fabricmc.net/docs/fabric-loader-0.11.3/net/fabricmc/loader/api/FabricLoader.html#isModLoaded(java.lang.String))

I hope this helps! Let me know if it wasn't quite what you were looking for.

**Reply by vlad2305m** [reply]
@thedocruby Nice. Thank you!

**Reply by thedocruby** [reply]
If you don't mind me asking, how did you get plasmo voice compatibility without knowing how to make mods compatible? PlasmoVoice GitHub clearly says that the mod is compatible with SoundPhysics. Did they do all the integration on their end?

**Reply by vlad2305m** [reply]
Yes. I pick up any properly registered vanilla sound source, so any mod is technically compatible if it uses vanilla sources.

**Reply by thedocruby** [reply]
@rght925h Apparently, Sound Physics Fabric is compatible with Plasmo-voice, a fork of Simple Voice Chat. Also, Sound Physics Remastered is a fork of Sound Physics Fabric, and is compatible with Simple Voice Chat, so you have some options.
